src/audio/pcm_converter/pcm_converter_generic.c

```diff
@@ -136,6 +138,63 @@ static void pcm_convert_s32_to_s24(const struct audio_stream *source,
 
 #endif /* CONFIG_FORMAT_S24LE && CONFIG_FORMAT_S32LE */
 
+#if CONFIG_FORMAT_FLOAT && (CONFIG_FORMAT_S16LE || CONFIG_FORMAT_S24LE || CONFIG_FORMAT_S32LE)
+
+static inline int32_t _pcm_shift(int32_t d, int32_t a)
+{
+	int64_t dd = d;
+
+	if (a > 32)
+		a = 32;
+	else if (a < -32)
+		a = -32;
+
+	dd = a >= 0 ? dd << a : dd >> -a;
+	if (dd > INT32_MAX)
+		dd = INT32_MAX;
+
+	return (int32_t)dd;
+}
+
+#define PCM_ABS32(x) (((x) ^ ((x) >> 31)) + (int)((unsigned int)(x) >> 31))
+
+static int32_t _pcm_convert_f_to_i(int32_t src)
+{
+	int32_t ex, mt, dst;
+
+	ex = (src >> 23);
+	ex = (ex & 0xFF) - 127; /* exponential */
+	mt = BIT(23) | (MASK(22, 0) & src); /* mantisa + 1.0 [Q9.22] */
+	/* calculate power */
+	dst = _pcm_shift(mt, ex - 23);
+	/* add 0.5 to round correctly but assert it doesn't lead to overflow */
+	if (ex - 22 < 9 || (src & BIT(31)) == BIT(31))
+		dst += _pcm_shift(mt, ex - 22) & 1;
+	/* copy sign to dst */
+	dst = (dst ^ (src >> 31)) + (int)((unsigned int)src >> 31);
+
+	return dst;
+}
+
+static int32_t _pcm_convert_i_to_f(int32_t src)
+{
+	int sign, mt, ex, dst, abs_clz;
+
+	if (src == 0)
+		return 0;
+
+	sign = src & BIT(31);
+	abs_clz = clz(PCM_ABS32(src));
+	ex = (127 + 31 - abs_clz) & 0xFF;
+	mt = PCM_ABS32(src);
+	mt = _pcm_shift(mt, 23 - 31 + abs_clz) & MASK(22, 0);
+	dst = sign | (ex << 23) | mt;
```

Comment:
please also put a comment here explaining the conversion e.g describing the float bit fields wold help everyone understand where all the magic numbers come from.  I would also use exponent and mantissa instead of ex and mt here (I had to think for a few seconds to mentally map these).

Reply:
added